### ISSUE TITLE
Invalid Credentials On Mass Files Upload

### ISSUE THREAD

**Issue by azaiter** [googleapis/nodejs-storage#439]
#### Environment details

  - OS: Linux 64bit, Running in docker containers on a GKE cluster.
  - Node.js version: v8.11.4
  - npm version: N/A (The one that comes with v8.11.4)
  - `@google-cloud/storage` version: ^2.0.3

#### Steps to reproduce

- We have almost 50 pods, and each pod is generating 3 files every X minutes. Once those files are generated, they all get uploaded at the same time to GCS. So in total, we have around 150 files that get uploaded in sequence (my application "on a pod" `await`s on each file of those three files to upload then start the next one, please note that we do have another 49 pods who are doing the same thing at the same time).
- I've noticed that a random %10 to %20 of those pods throw a 401 error that looks like the following:
```
{
    "code": 401,
    "errors": [
        {
            "domain": "global",
            "reason": "authError",
            "message": "Invalid Credentials",
            "locationType": "header",
            "location": "Authorization"
        }
    ],
    "message": "Invalid Credentials"
}
```

Please also note that this error only occurs on the FIRST file that gets uploaded and not the following 2 files. I have autoRetry on, and I end up seeing the "failed first" file after few seconds on GCS.

So if you want to replicate this, you could have something that does the following:
> Spawn large N number of node pods or processes that uploads a file on a medium interval (like 10-15 minutes) at the same time.

Could it be that I might get rate limited? if so, is there any documentation about that? Then if so, why wouldn't the subsequent 2 files throw the same error?

Thanks!

Update # 1: after testing more, the errors start to appear after 30-60 minutes of pods operation. So 900-1500 files get uploaded before those errors show up. 
Update # 2: This issues existed since 1.6+ versions, in 1.6+ I had ECONNRESET in addition to the above 401 but that was a user error because I was uploading all three files at the same time and not utilizing async/await operations.
Update # 3 : the number of pods is around 50, and not 100

**Comment by azaiter** [comment]
I enabled cloud storage access Logs & storage Logs in order to further investigate the issue from our side as well. We have noticed that the log entry for the files that throw 401 status code has:
- 122 cs_bytes the number of bytes sent in the request. 
- 249 sc_bytes the number of bytes sent in the response.
- An empty cs_object specified in this request, the rest of the successful calls has it defined.

**Comment by azaiter** [comment]
After further deep investigation and creating a binary written a different language (go) to identify and have a clear understanding of this issue, we have finally found the root cause of this error from the NodeJS side and our Kubernetes side and we were able to address both of them accordingly based on the resolution that I'll be explaining later in this response.
The go binary had a very similar behavior but was throwing a different error, the error produced when this incident happens is:
```
Failed to close file write stream: Post https://www.googleapis.com/upload/storage/v1/b/<BUCKET_NAME>/o?alt=json&prettyPrint=false&projection=full&uploadType=multipart: oauth2: cannot fetch token: Post https://accounts.google.com/o/oauth2/token: net/http: TLS handshake timeout
```
By looking at the error, I immediately rolled out NodeJS and Go applications out of the equation and started to dig deep into why the oauth2 token requests were having TLS handshake timeout. By looking into a similar closed issue that Kubernetes project has [here](https://github.com/kubernetes/kubernetes/issues/56293). I discovered that random [CPU starvation was happening before writing the objects to GCS](https://github.com/kubernetes/kubernetes/pull/56394#issuecomment-361558277) on the pods as we were resource limiting them. To confirm this theory, I decided to limit them by a margin and most of the pods failed with the same error as a result. __The 401 on the NodeJS error was kind of misleading and wasn't pointing at this issue by any means__, but the Go alternative was the key for finding this issue and resolving it. Given that we don't encounter this issue on most of the pods, and exponential backoff was the way to resolve the issue by the kubernetes team, and since we have the __"autoRety" flag enabled that does the exponential backoff__. We've decided to keep the code as is and consider the issue as __"resolved with an explanation - no actions needed."__
